src/PythonProjectBootstrapper/ProjectGenerationUtils.py

```diff
@@ -0,0 +1,204 @@
+# ----------------------------------------------------------------------
+# |
+# |  Copyright (c) 2024 Scientific Software Engineering Center at Georgia Tech
+# |  Distributed under the MIT License.
+# |
+# ----------------------------------------------------------------------
+import hashlib
+import itertools
+import os
+import sys
+from pathlib import Path
+
+from rich import print  # pylint: disable=redefined-builtin
+from rich.panel import Panel
+import yaml
+
+from dbrownell_Common import PathEx
+from PythonProjectBootstrapper import __version__
+
+# The following imports are used in cookiecutter hooks. Import them here to
+# ensure that they are frozen when creating binaries,
+import shutil  # pylint: disable=unused-import, wrong-import-order
+import textwrap  # pylint: disable=unused-import, wrong-import-order
+
+
+HASH_ALG = hashlib.sha256
+
+
+# ----------------------------------------------------------------------
+def _GenerateFileHash(filepath: Path, hash_fn) -> str:
+    PathEx.EnsureFile(filepath)
+    with open(filepath, "rb") as file:
+        hash_value = hashlib.file_digest(file, hash_fn).hexdigest()
+
+    return hash_value
+
+
+# ----------------------------------------------------------------------
+def _CreateManifest(generated_dir: Path) -> dict[str, str]:
+    manifest_dict: dict[str, str] = {}
+    generated_files: list[Path] = []
+
+    for root, _, files in os.walk(generated_dir):
+        generated_files += [Path(root) / Path(file) for file in files]
+
+    # Populate manifest dictionary
+    for genfile in generated_files:
+        PathEx.EnsureFile(genfile)
+        rel_path = PathEx.CreateRelativePath(generated_dir, genfile)
+        manifest_dict[str(rel_path)] = _GenerateFileHash(filepath=genfile, hash_fn=HASH_ALG)
+
+    return manifest_dict
+
+
+# ----------------------------------------------------------------------
+def _ConditionallyRemoveUnchangedTemplateFiles(
+    new_manifest_dict: dict[str, str],
+    existing_manifest_dict: dict[str, str],
+    output_dir: Path,
+) -> None:
+    # Removes any template files no longer being generated as long as the file was never modified by the user
+
+    # files no longer in template
+    removed_template_files: set[Path] = set(existing_manifest_dict.keys()) - set(
+        new_manifest_dict.keys()
+    )
+
+    PathEx.EnsureDir(output_dir)
+
+    # remove files no longer in template if they are unchanged
+    for removed_file_rel_path in removed_template_files:
+        removed_full_path = output_dir / removed_file_rel_path
+
+        if removed_full_path.is_file():
+            current_hash = _GenerateFileHash(filepath=removed_full_path, hash_fn=HASH_ALG)
+            original_hash = existing_manifest_dict[removed_file_rel_path]
+
+            if current_hash == original_hash:
+                removed_full_path.unlink()
+
+
+# ----------------------------------------------------------------------
+def _CopyToOutputDir(
```

Comment:
Rename to `CopyToOutputDir` because it is used outside of this file.

Reply:
Done